src/rust/ide/view/graph-editor/src/component/node/action_bar.rs

```diff
@@ -0,0 +1,240 @@
+            let rect_rounded         = rect.corners_radius(corner_radius);
+            let rect_filled          = rect_rounded.fill(HOVER_COLOR);
+            rect_filled.into()
+        }
+    }
+}
+
+
+
+// ===========
+// === Frp ===
+// ===========
+
+ensogl::define_endpoints! {
+    Input {
+        set_size   (Vector2),
+        show_icons (),
+        hide_icons (),
+    }
+    Output {
+        mouse_over       (),
+        mouse_out        (),
+        action_visbility (bool),
+        action_freeze    (bool),
+        action_skip      (bool),
+    }
+}
+
+
+
+// ========================
+// === Action Bar Model ===
+// ========================
+
+#[derive(Clone,CloneRef,Debug)]
+struct Model {
+    hover_area            : component::ShapeView<hover_rect::Shape>,
+
+    icons                 : display::object::Instance,
+    icon_freeze           : ToggleButton<node::icon::action::freeze::Shape>,
+    icon_visibility       : ToggleButton<node::icon::action::visibility::Shape>,
+    icon_skip             : ToggleButton<node::icon::action::skip::Shape>,
+
+    display_object        : display::object::Instance,
+    size                  : Rc<Cell<Vector2>>,
+
+    all_shapes            : compound_shape::Events,
+
+}
+
+impl Model {
+    fn new(app:&Application) -> Self {
+        let scene                 = app.display.scene();
+        let logger                = Logger::new("ActionBarModel");
+        let hover_area            = component::ShapeView::new(&logger,scene);
+        let icon_freeze           = ToggleButton::new(&app);
+        let icon_visibility       = ToggleButton::new(&app);
+        let icon_skip             = ToggleButton::new(&app);
+        let all_shapes            = compound_shape::Events::default();
+
+        all_shapes.add_sub_shape(&hover_area);
+        all_shapes.add_sub_shape(&icon_freeze.view());
+        all_shapes.add_sub_shape(&icon_visibility.view());
+        all_shapes.add_sub_shape(&icon_skip.view());
+
+        let display_object        = display::object::Instance::new(&logger);
+        let icons                 = display::object::Instance::new(&logger);
+
+        let size                  = default();
+
+        Self{hover_area,icons,display_object,size,icon_freeze,icon_visibility,
+             icon_skip,all_shapes}.init()
+    }
+
+    fn init(self) -> Self {
+        self.add_child(&self.hover_area);
+
+        self.add_child(&self.icons);
+        self.icons.add_child(&self.icon_freeze);
+        self.icons.add_child(&self.icon_skip);
+        self.icons.add_child(&self.icon_visibility);
+
+        // Default state s hidden.
+        self.hide();
+        self
+    }
+
+    fn place_button_in_slot<T:ColorableShape>(&self, button:&ToggleButton<T>, index:usize) {
+        let icon_size = Vector2::new(self.size.get().y, self.size.get().y);
+        let index     = index as f32;
+        button.mod_position(|p| p.x = (1.5 * index + 0.5) * icon_size.x);
```

Comment:
these numbers are magical. Please extract them at least to well named variables :)